Common/SensorFusion/Src/SensorFusion.cpp

```diff
@@ -0,0 +1,636 @@
+
+    float IMinKMultH[DIM*DIM];
+    for (int i = 0; i < DIM*DIM; i++) IMinKMultH[i] = I[i] - kMultH[i];
+
+    float newP[DIM*DIM];
+    mul(IMinKMultH, p, newP, DIM, DIM, DIM);
+
+
+    /*Output*/
+
+    float northSpeed = x[3], eastSpeed = x[5];
+    float track = RAD_TO_DEG(atan2(eastSpeed, northSpeed));
+    if (track < 0) track += 360;
+
+    float groundSpeed = sqrt(northSpeed*northSpeed + eastSpeed*eastSpeed);
+
+#ifdef AUTOPILOT
+    double * latLongOut = cartesianToGPS(x[2], x[4]);
+#else
+    double latLongOut[2] = {x[2], x[4]};
+#endif
+
+    Output->altitude = x[0];
+    Output->rateOfClimb = x[1];
+    Output->latitude = latLongOut[0];
+    Output->latitudeSpeed = x[3];
+    Output->longitude = latLongOut[1];
+    Output->longitudeSpeed = x[5];
+    Output->track = track;
+    Output->groundSpeed = groundSpeed;
+
+    for (int i = 0; i < DIM*1; i++) iterdata->prevX[i] = newX[i];
+    for (int i = 0; i < DIM*DIM; i++) iterdata->prevP[i] = newP[i];
+
+    return SFError;
+}
+
+SFError_t SF_GenerateNewResult()
+{
+    SFError_t SFError;
+    SFError.errorCode = 0;
+    
+    IMUData_t imuData;
+    GpsData_t GpsData;
+    AltimeterData_t altimeterData;
+    airspeedData_t airspeedData;
+    imuObj->GetResult(imuData);
+#ifdef AUTOPILOT
+    gpsObj->GetResult(GpsData);
+    altimeterObj->GetResult(altimeterData);
+    airspeedObj->GetResult(airspeedData);
+
+    //Send gps timestamp
+    #ifndef UNIT_TESTING
+    TimeStampingPOGI *timeStamper = TimeStampingPOGI::GetInstance();
+
+    if(GpsData.ggaDataIsNew)
+    {
+        timeStamper->setGPSTime(&GpsData);
+    }
+    #endif
+#endif
+
+    SFAttitudeOutput_t attitudeOutput;
+    SFPathOutput_t pathOutput;
+
+    SFError = SF_GetAttitude(&attitudeOutput, &imuData);
+    if(SFError.errorCode != 0) return SFError;
+
+    SFError = SF_GetPosition(&pathOutput, &altimeterData, &GpsData, &imuData, &attitudeOutput, &iterData);
+    if(SFError.errorCode != 0) return SFError;
+
+    SFOutput.pitch = attitudeOutput.pitch;
+    SFOutput.roll = attitudeOutput.roll;
+    SFOutput.yaw = attitudeOutput.yaw;
+    SFOutput.pitchRate = attitudeOutput.pitchRate;
+    SFOutput.rollRate = attitudeOutput.rollRate;
+    SFOutput.yawRate = attitudeOutput.yawRate;
+    SFOutput.altitude = pathOutput.altitude;
+    SFOutput.rateOfClimb = pathOutput.rateOfClimb;
+    SFOutput.latitude = pathOutput.latitude;
+    SFOutput.latitudeSpeed = pathOutput.latitudeSpeed;
+    SFOutput.longitude = pathOutput.longitude;
+    SFOutput.longitudeSpeed = pathOutput.longitudeSpeed;
+    SFOutput.track = pathOutput.track;
+    SFOutput.groundSpeed = pathOutput.groundSpeed;
+    SFOutput.heading = attitudeOutput.heading;
+
+    return SFError;
+}
+
+SFError_t SF_GetResult(SFOutput_t *output)
+{
+    SFError_t SFError;
+    SFError.errorCode = 0;
+
+    *output = SFOutput;
+
+    return SFError;
+}
+
+IMU_Data_t SF_GetRawIMU()
+{
+    IMUData_t imuData;
+    imuObj->GetResult(imuData);
```

Comment:
did you check if this is workable with Neha's new IMU driver?

Reply:
Yes, it inherits the IMU class so it works. Though for target builds, the exact model of each sensor will need to be specified in SF_init().

Reply:
Sounds good, I think we may have to rename MPUO_6050 driver to imu.cpp and imu.hpp since that's what seems to be called